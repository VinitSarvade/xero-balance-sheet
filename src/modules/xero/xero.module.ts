export type {
  BalanceSheetReport,
  BalanceSheetRow,
  HeaderRow,
  SectionRow,
  SectionData,
  Cell,
} from "./xeroReport.dto";
export { getReport } from "./xeroReport.service";
