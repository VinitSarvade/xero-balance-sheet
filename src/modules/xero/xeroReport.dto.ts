export interface BalanceSheetReport {
  ReportID: string;
  ReportName: string;
  ReportTitles: string[];
  ReportDate: string;
  Rows: BalanceSheetRow[];
}

export type BalanceSheetRow = SectionRow | HeaderRow;

export interface HeaderRow {
  RowType: "Header";
  Cells: Cell[];
}

export interface Cell {
  Value: string;
}

export interface SectionRow {
  RowType: "Section";
  Title: string;
  Rows: SectionData[];
}

export type SectionData = DataRow | SummaryRow;

export interface DataRow {
  RowType: "Row";
  Cells: Cell[];
}

export interface SummaryRow {
  RowType: "SummaryRow";
  Cells: Cell[];
}
