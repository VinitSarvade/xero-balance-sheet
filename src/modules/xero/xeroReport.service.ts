import { BalanceSheetReport } from "./xeroReport.dto";

const BALANCE_SHEET_API = `${process.env.XERO_API_BASE_URL}/api.xro/2.0/Reports/BalanceSheet`;

export async function getReport(): Promise<BalanceSheetReport> {
  const response = await fetch(BALANCE_SHEET_API, {
    headers: {
      // assuming authentication is done by API token
      Authorization: `Bearer ${process.env.XERO_API_TOKEN ?? ""}`,
    },
    cache: "force-cache",
  });

  if (!response.ok) {
    throw new Error(response.statusText);
  }

  const result = await response.json();

  return result.Reports[0];
}
