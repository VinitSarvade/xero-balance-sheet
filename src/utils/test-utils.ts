import {
  format as prettyFormat,
  plugins as prettyFormatPlugins,
} from "pretty-format";
import { http, HttpResponse } from "msw";
import { BalanceSheetReport } from "@/modules/xero/xeroReport.dto";

export function format(value: any) {
  return prettyFormat(value, {
    escapeRegex: true,
    indent: 2,
    plugins: [
      prettyFormatPlugins.ReactTestComponent,
      prettyFormatPlugins.ReactElement,
      prettyFormatPlugins.DOMElement,
      prettyFormatPlugins.DOMCollection,
      prettyFormatPlugins.Immutable,
      prettyFormatPlugins.AsymmetricMatcher,
    ],
    printFunctionName: false,
  });
}

export const BalanceSheetMock = {
  ReportID: "report-id",
  ReportName: "Balance Sheet",
  ReportDate: "2022-01-01",
  ReportTitles: ["Balance Sheet", "Test Company"],
  Rows: [
    {
      RowType: "Header",
      Cells: [{ Value: "" }, { Value: "2023" }, { Value: "2024" }],
    },
    {
      RowType: "Section",
      Title: "Assets",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Bank",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "My Bank Account" },
            { Value: "126.70" },
            { Value: "99.60" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business" },
            { Value: "92911.00" },
            { Value: "92911.00" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business 1" },
            { Value: "11039.00" },
            { Value: "11039.00" },
          ],
        },
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Bank" },
            { Value: "104076.70" },
            { Value: "104049.60" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "Liabilities",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Current Liabilities",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "Car Loan" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "",
      Rows: [
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Current Liabilities" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
  ],
} as BalanceSheetReport;

export function getHandlers() {
  return [
    http.get(`${process.env.BASE_URL}/api/xero/balance-sheet`, () => {
      return HttpResponse.json(BalanceSheetMock);
    }),
  ];
}
