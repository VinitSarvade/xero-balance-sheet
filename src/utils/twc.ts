import { createTwc } from "react-twc";
import { cn } from "./cn";

export const twc = createTwc({ compose: cn });
