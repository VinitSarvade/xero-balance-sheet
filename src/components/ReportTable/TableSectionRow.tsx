import { SectionData, SectionRow } from "@/modules/xero/xero.module";
import { twc } from "@/utils/twc";
import TableSectionColumn from "./TableSectionColumn";

interface TableSectionRowProps {
  row: SectionData;
  section: SectionRow;
}

export const PrimarySectionSummaryRow = twc.tr`bg-sky-100 text-lg font-semibold uppercase tracking-tight`;
export const SectionSummaryRow = twc.tr`border-solid border-gray-400 border-y bg-sky-100`;
export const NetRow = twc.tr`bg-orange-100 text-lg font-semibold uppercase tracking-tight`;
export const Row = twc.tr``;

type RowComponentType =
  | typeof PrimarySectionSummaryRow
  | typeof SectionSummaryRow
  | typeof NetRow
  | typeof Row;

export function getRowComponent(
  row: SectionData,
  section: SectionRow,
): [
  RowComponentType,
  { isPrimarySummaryRow?: boolean; isSummaryRow?: boolean; isNetRow?: boolean },
] {
  switch (row.RowType) {
    case "SummaryRow":
      if (section.Title.length === 0 && section.Rows.length === 1) {
        return [PrimarySectionSummaryRow, { isPrimarySummaryRow: true }];
      }
      return [SectionSummaryRow, { isSummaryRow: true }];

    case "Row":
      if (row.Cells[0].Value === "Net Assets") {
        return [NetRow, { isNetRow: true }];
      }

      return [Row, {}];
  }
}

export default function TableSectionRow({
  section,
  row,
}: TableSectionRowProps) {
  const [RowComponent, cellProps] = getRowComponent(row, section);

  return (
    <RowComponent>
      {row.Cells.map((cell, idx) => (
        <TableSectionColumn
          key={`cell-${idx}-${section.Title}-${cell.Value}`}
          cell={cell}
          particularsColumn={idx === 0}
          amountColumn={idx > 0}
          {...cellProps}
        />
      ))}
    </RowComponent>
  );
}
