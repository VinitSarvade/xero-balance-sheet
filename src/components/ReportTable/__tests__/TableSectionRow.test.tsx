import { expect, describe, it } from "bun:test";
import { render } from "@testing-library/react";
import { SectionData, SectionRow } from "@/modules/xero/xero.module";
import { format } from "@/utils/test-utils";
import TableSectionRow, {
  getRowComponent,
  PrimarySectionSummaryRow,
  SectionSummaryRow,
  NetRow,
  Row,
} from "../TableSectionRow";

describe("getRowComponent", () => {
  it("returns PrimarySectionSummaryRow component for primary summary row", () => {
    const row: SectionData = { RowType: "SummaryRow", Cells: [] };
    const section: SectionRow = {
      Title: "",
      Rows: [row],
      RowType: "Section",
    };
    const [component, props] = getRowComponent(row, section);
    expect(component).toBe(PrimarySectionSummaryRow);
    expect(props).toEqual({ isPrimarySummaryRow: true });
  });

  it("returns SectionSummaryRow component for non-primary summary row", () => {
    const row: SectionData = { RowType: "SummaryRow", Cells: [] };
    const section: SectionRow = {
      Title: "Section Title",
      Rows: [row],
      RowType: "Section",
    };
    const [component, props] = getRowComponent(row, section);
    expect(component).toBe(SectionSummaryRow);
    expect(props).toEqual({ isSummaryRow: true });
  });

  it("returns NetRow component for row with 'Net Assets' cell value", () => {
    const row: SectionData = {
      RowType: "Row",
      Cells: [{ Value: "Net Assets" }],
    };
    const section: SectionRow = {
      Title: "",
      Rows: [row],
      RowType: "Section",
    };
    const [component, props] = getRowComponent(row, section);
    expect(component).toBe(NetRow);
    expect(props).toEqual({ isNetRow: true });
  });

  it("returns Row component for regular row", () => {
    const row: SectionData = {
      RowType: "Row",
      Cells: [{ Value: "Regular Row" }],
    };
    const section: SectionRow = {
      Title: "",
      Rows: [row],
      RowType: "Section",
    };
    const [component, props] = getRowComponent(row, section);
    expect(component).toBe(Row);
    expect(props).toEqual({});
  });
});

describe("TableSectionRow", () => {
  it("renders the correct number of TableSectionColumn components", () => {
    const section: SectionRow = {
      Title: "Section Title",
      Rows: [],
      RowType: "Section",
    };
    const row: SectionData = {
      RowType: "Row",
      Cells: [{ Value: "Cell 1" }, { Value: "Cell 2" }, { Value: "Cell 3" }],
    };
    const { getAllByRole } = render(
      <table>
        <tbody>
          <TableSectionRow section={section} row={row} />
        </tbody>
      </table>,
    );
    const columns = getAllByRole("cell");
    expect(columns).toHaveLength(3);
    expect(columns[0].textContent).toBe("Cell 1");
    expect(columns[1].textContent).toBe("Cell 2");
    expect(columns[2].textContent).toBe("Cell 3");
  });

  it("matches the snapshot", () => {
    const section: SectionRow = {
      Title: "Section Title",
      Rows: [],
      RowType: "Section",
    };
    const row: SectionData = {
      RowType: "Row",
      Cells: [{ Value: "Cell 1" }, { Value: "Cell 2" }, { Value: "Cell 3" }],
    };
    const { asFragment } = render(
      <table>
        <tbody>
          <TableSectionRow section={section} row={row} />
        </tbody>
      </table>,
    );

    expect(format(asFragment())).toMatchSnapshot();
  });

  it("renders the correct RowComponent for SummaryRow", () => {
    const section: SectionRow = {
      Title: "Section Title",
      Rows: [],
      RowType: "Section",
    };
    const summaryRow: SectionData = {
      RowType: "SummaryRow",
      Cells: [{ Value: "Summary Row" }],
    };

    const { getByRole } = render(
      <table>
        <tbody>
          <TableSectionRow section={section} row={summaryRow} />
        </tbody>
      </table>,
    );
    expect(getByRole("row").textContent).toBe("Summary Row");
  });

  it("renders the correct RowComponent for Row", () => {
    const section: SectionRow = {
      Title: "Section Title",
      Rows: [],
      RowType: "Section",
    };
    const regularRow: SectionData = {
      RowType: "Row",
      Cells: [{ Value: "Regular Row" }],
    };
    const { getByRole } = render(
      <table>
        <tbody>
          <TableSectionRow section={section} row={regularRow} />
        </tbody>
      </table>,
    );
    expect(getByRole("row").classList).toHaveLength(0);
  });
});
