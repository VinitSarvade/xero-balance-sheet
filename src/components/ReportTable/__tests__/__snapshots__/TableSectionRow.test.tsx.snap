// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSectionRow matches the snapshot 1`] = `
"<DocumentFragment>
  <table>
    <tbody>
      <tr
        class=""
      >
        <td
          class="px-2 py-1.5 pl-4"
        >
          Cell 1
        </td>
        <td
          class="px-2 text-right"
        >
          Cell 2
        </td>
        <td
          class="px-2 text-right"
        >
          Cell 3
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>"
`;
